### PR TITLE
feat(nargo): Add `--exact` flag to `nargo test`

### LINKED ISSUE #2271: The `nargo test` command should have a way to select exactly one test
### Problem

In #2017, the module path was prefixed to test names; however, if you run a test in the root package, such as `test_main`, it would run any `test_main` function in any submodule since their full file path contains `test_main`.

This could be considered a bug, but it requires a new feature. Cargo has the `--exact` flag which does full equality instead of a string match.

### Happy Case

Add `--exact` flag and use equality check instead of `.contains()`

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by phated**
# Description

## Problem\*

Resolves #2271  

## Summary\*

This adds the `--exact` flag to `nargo test` which allows us to select tests based on an exact name. I also refactored the code to make it more clear which tests we were selecting and fixed problems with running tests in the vscode plugin inside a workspace (relies on https://github.com/noir-lang/vscode-noir/pull/29)

## Documentation

- [x] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [x] I will request for and support Dev Rel's help in documenting this PR.

  
  

The `--exact` flag needs to be documented on the `nargo test` command.

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_frontend/src/hir/mod.rs [hunk 1]
```diff
@@ -76,28 +102,28 @@ impl Context {
...
-                path.contains(pattern).then_some((path, id))
+                let fully_qualified_name = self.fully_qualified_function_name(crate_id, &id);
+                match &pattern {
+                    FunctionNameMatch::Anything => Some((fully_qualified_name, id)),
+                    FunctionNameMatch::Exact(pattern) => {
+                        if &fully_qualified_name == pattern {
+                            Some((fully_qualified_name, id))
+                        } else {
+                            None
+                        }
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir/mod.rs:121]:
```suggestion
                        (&fully_qualified_name == pattern).then_some((fully_qualified_name, id))
```

File: crates/noirc_frontend/src/hir/mod.rs [hunk 2]
```diff
@@ -52,10 +59,29 @@ impl Context {
...
     pub fn function_name(&self, id: &FuncId) -> &str {
         self.def_interner.function_name(id)
     }
 
+    pub fn fully_qualified_function_name(&self, crate_id: &CrateId, id: &FuncId) -> String {
+        let def_map = self.def_map(crate_id).expect("The local crate should be analyzed already");
+
+        let name = self.def_interner.function_name(id);
+
+        let meta = self.def_interner.function_meta(&id);
```
Comment (TomAFrench) [inline crates/noirc_frontend/src/hir/mod.rs:72]:
```suggestion
        let meta = self.def_interner.function_meta(id);
```

File: crates/noirc_frontend/src/hir/mod.rs (hunk 1 above)
Comment (phated) [reply crates/noirc_frontend/src/hir/mod.rs:121]: TIL! I figured that `contains` returned an option and this was an available function, didn't realize it was implemented on bools 🎉

**Comment by Savio-Sou** [comment]
Is this going into v0.10.0? Seems that it wasn't automatically added into https://github.com/noir-lang/noir/pull/2039?

**Comment by kevaundray** [comment]
> Is this going into v0.10.0? Seems that it wasn't automatically added into #2039?

#2039 is only updated when something is merged into master, its represents the current version that we have not released yet

**Review by kobyhallx — APPROVED** [review]

**Pull request merged**